src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -139,12 +155,17 @@ public Workspace Workspace
     /// </summary>
     public Project? GetProject(ProjectId? projectId)
     {
-        if (this.ContainsProject(projectId))
+        if (projectId is null || !_projectIdToIndex.TryGetValue(projectId, out var projectIndex))
+            return null;
+
+        var project = _projects[projectIndex];
+        if (project == null)
         {
-            return ImmutableInterlocked.GetOrAdd(ref _projectIdToProjectMap, projectId, s_createProjectFunction, this);
+            project = CreateProject(projectId, this);
+            Interlocked.CompareExchange(ref _projects[projectIndex], project, null);
         }
 
-        return null;
+        return project;
```

Comment:
this is wrong/racey.  you are returning the project that might have *failed* to go into the dictionary.  You need to erad back from `_projects` again to get the project if another thread beats you.